ext/RMagick/rmagick.h

```diff
@@ -973,6 +1000,15 @@ extern VALUE Image_wave(int, VALUE *, VALUE);
 extern VALUE Image_wet_floor(int, VALUE *, VALUE);
 extern VALUE Image_white_threshold(int, VALUE *, VALUE);
 extern VALUE Image_write(VALUE, VALUE);
+#if defined(IMAGEMAGICK_7)
+extern VALUE Image_composite_mask(int, VALUE *, VALUE);
+extern VALUE Image_read_mask(int, VALUE *, VALUE);
+extern VALUE Image_write_mask(int, VALUE *, VALUE);
```

Comment:
Can you explain some how these methods are different from the ones below?

Reply:
The masking is done differently in ImageMagick 7. Instead of using an image as a mask it is now part of the new channel structure. Each pixel in ImageMagick has a set of channels instead of always 5 inside ImageMagick 6. This can be one channel (Gray) or 5 (CMYKA) or maybe even 30 (???) . This depends on the image. These methods set a channel inside the pixel and they are used in other methods. If you have an RGB image (3 channels) and you set the read mask this adds an extra channel to determine if the pixel can be read in other operations.

Reply:
Wow, thanks for the great explanation! I'm trying to think if there's a way to smoothly transition across both. Would it be possible to simulate the IM6 behavior compiled against IM7? Like, could we define a version of `Image_combine()` that uses IM7 features to produce the same result?

Reply:
The behavior is different in IM6 and IM7 and I don't know if you could simulate the IM6 behavior in IM7. The composite version might do that but it could also be that you need to set multiple tasks. And then you would lose the extra feature of those separate mask.

New features will (and have) been added to ImageMagick 7. When we will add those here they will only work for IM7 and not be available in IM6. And that is also why I keep advising to distribute two different versions of RMagick.

Reply:
> And then you would lose the extra feature of those separate mask.

I think that is fine, since we're trying to emulate the behavior of IM6 for consistency. I think we can add new methods that take advantage of IM7 functionality, and also maintain the existing features for the time being. So in this case, maybe we could have:

```
VALUE Image_composite_mask(int, VALUE *, VALUE) {
#if defined(IMAGEMAGICK_7)
  // do the thing
#else
  rb_raise("method only available in IM 7");
#endif
}

VALUE Image_combine(int, VALUE *, VALUE) {
#if defined(IMAGEMAGICK_7)
  // apply extra masks
#else
  // do the thing
#endif

}
```

Reply:
I am against methods that raise a "not supported" runtime exception. Doing a hard split can avoid this. I really think it is more clear to the users than these kind of hacks. Maybe we should do the same thing as we have done in the ImageMagick project and write our own porting guide that contains the relevant information from this guide: https://imagemagick.org/script/porting.php.

And `Image_combine` uses a colorspace instead of channels now and this has been moved to the `ImageList` class where it makes more sense to have this method. We could also do that for the IM6 build and then depending on the size of the list the channels would be determined. But you still cannot use a colorspace argument because that only works for ImageMagick 7.

Reply:
> I am against methods that raise a "not supported" runtime exception.

Can you talk more about why you don't like this? I agree it's not ideal, but I think it makes the transition path a lot more smooth for our users, so I'm more okay with it. I think the real ideal would be if we could emulate IM7 behavior against IM6, but that seems much harder. The way I see it, we're adding functionality, so we're not breaking anything existing. I suppose another alternative would be to not even define the method for IM6, but then the user might be confused about why it's not there when they find it in the docs somewhere.

Reply:
New methods that don't work in IM6 will have been used in the "old" version so there is not really a transition path. And then when someone tries to call that method and finds out it is "not supported" will also not be very nice for the end user. And it think it would be even more confusing if it would be in the docs but wouldn't work when linked with IM6.

I think that the transition path for most users will already be very smooth. If you take a look at the differences in the API in this PR the set of methods that has changed is very small. Only the ones that have a really different behavior have been changed.

Reply:
After a chat we came to the conclusion that it would be possible to make `mask` backwards compatible. Will update this PR accordingly. And deprecating `combine` and moving it towards the `ImageList` should be moved to a separate PR first.